File: backend/app/schemas/manual_entry.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import date
from uuid import UUID
from decimal import Decimal


class ManualEntryShot(BaseModel):
    shot_number: str
    side: Optional[str] = None
    vest_number: Optional[str] = None
    angle_degrees: Optional[Decimal] = None
    caliber: Optional[str] = None
    velocity_m_s: Optional[Decimal] = None
    trauma_mm: Optional[Decimal] = None
    trauma_qualitative: Optional[str] = None
    protection_level: Optional[str] = None
    temperature_c: Optional[Decimal] = None
    humidity_percent: Optional[Decimal] = None


class ManualEntryVestTab(BaseModel):
    vest_number: Optional[str] = None
    size: Optional[str] = None
    conditioning: Optional[str] = None
    ballistic_limit: Optional[bool] = False
    shots: List[ManualEntryShot] = []


class ManualEntryRequest(BaseModel):
    name: str
    test_date: Optional[date] = None
    lab_name: Optional[str] = None
    protocol: Optional[str] = None
    clay_temperature_c: Optional[Decimal] = None
    ambient_temperature_c: Optional[Decimal] = None
    humidity_percent: Optional[Decimal] = None
    vest_id: Optional[str] = None
    geometry_id: str
    is_official: Optional[bool] = False
    certification_number: Optional[str] = None
    notes: Optional[str] = None
    protection_level: Optional[str] = None
    vest_tabs: List[ManualEntryVestTab] = []


class ManualEntryResponse(BaseModel):
    parent_session_id: str
    child_session_ids: List[str] = []
    total_shots: int

File: frontend/src/api/manualEntry.ts
import { apiClient } from './client';

export interface ManualEntryShot {
  shot_number: string;
  side?: string | null;
  conditioning?: string | null;
  vest_number?: string | null;
  angle_degrees?: number | null;
  caliber?: string | null;
  velocity_m_s?: number | null;
  trauma_mm?: number | null;
  trauma_qualitative?: string | null;
  protection_level?: string | null;
  temperature_c?: number | null;
  humidity_percent?: number | null;
}

export interface ManualEntryVestTab {
  vest_number?: string | null;
  serial_number?: string | null;
  size?: string | null;
  conditioning?: string | null;
  ballistic_limit?: boolean | false;
  shots: ManualEntryShot[];
}

export interface ManualEntryRequest {
  name: string;
  test_date?: string | null;
  lab_name?: string | null;
  protocol?: string | null;
  clay_temperature_c?: number | null;
  ambient_temperature_c?: number | null;
  humidity_percent?: number | null;
  vest_id?: string | null;
  geometry_id: string;
  is_official?: boolean | false;
  certification_number?: string | null;
  notes?: string | null;
  protection_level?: string | null;
  vest_tabs: ManualEntryVestTab[];
}

export interface ManualEntryResponse {
  parent_session_id: string;
  child_session_ids: string[];
  total_shots: number;
}

export const manualEntryApi = {
  submit: (entry: ManualEntryRequest) =>
    apiClient.post<ManualEntryResponse>('/api/v1/test-sessions/manual-entry', entry),
};
